File: src/j2ee/com.ibm.sbt.web/WebContent/js/sdk/sbt/base/JsonDataHandler.js
/*
 * © Copyright IBM Corp. 2012,2013
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License. 
 * You may obtain a copy of the License at:
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, 
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
 * implied. See the License for the specific language governing 
 * permissions and limitations under the License.
 */

/**
 * Social Business Toolkit SDK. 
 * Helpers for the base capabilities of data handlers.
 * 
 * @module sbt.base.JsonDataHandler
 */
define(["../declare", "../lang", "../json", "./DataHandler", "../Jsonpath", "../stringUtil"], 
    function(declare, lang, json, DataHandler, jsonPath, stringUtil) {

    /**
     * JsonDataHandler class
     * 
     * @class JsonDataHandler
     * @namespace sbt.base
     */     
	var JsonDataHandler = declare(DataHandler, {	
		
        /**
         * Data type for this DataHandler is 'json'
         */
        dataType : "json",
		
        /**
         * Set of jsonpath expressions used by this handler.
         */
        jsonpath : null,
        
        /**
         * Set of values that have already been read.
         */
        _values : null,
        
	    /**
	     * @constructor
	     * @param {Object} args Arguments for this data handler.
	     */     
		constructor : function(args) {
            lang.mixin(this, args);
            this._values = {}; 
            this.data = args.data;
		},
        
        /**
         * @method getAsString
         * @param data
         * @returns
         */
        getAsString : function(property) {
        	this._validateProperty(property, "getAsString");
        	if (this._values) {
                if (!this._values.hasOwnProperty(property)) {
                    this._values[property] = this._get(property).toString();
                }
                return this._values[property];
            } else {
                return _get(property);
            }
        },

        /**
         * @method getAsNumber
         * @returns
         */
        getAsNumber : function(property) {
        	this._validateProperty(property, "getAsNumber");
        	if (this._values) {
                if (!this._values.hasOwnProperty(property)) {
                    this._values[property] = this._getNumber(property); 
                }
                return this._values[property];
            } else {
                return _getNumber(property);
            }
        },

        /**
         * @method getAsDate
         * @returns
         */
        getAsDate : function(property) {
        	this._validateProperty(property, "getAsDate");
        	if (this._values) {
                if (!this._values.hasOwnProperty(property)) {
                    this._values[property] = this._getDate(property); 
                }
                return this._values[property];
            } else {
                return _getDate(property);
            }
        },

        /**
         * @method getAsBoolean
         * @returns
         */
        getAsBoolean : function(property) {
        	this._validateProperty(property, "getAsBoolean");
        	if (this._values) {
                if (!this._values.hasOwnProperty(property)) {
                    this._values[property] = this._getBoolean(property); 
                }
                return this._values[property];
            } else {
                return _getBoolean(property);
            }                    
        },

        /**
         * @method getAsArray
         * @returns
         */
        getAsArray : function(property) {
        	this._validateProperty(property, "getAsArray");
        	if (this._values) {
                if (!this._values.hasOwnProperty(property)) {
                    this._values[property] = this._get(property);
                }
                return this._values[property];
            } else {
                return _get(property);
            }       
        },

        /**
         * @method getEntityId
         * @returns
         */
        getEntityId : function(data) {
        	return stringUtil.trim(this.getAsString(this.jsonpath["id"]));
        },

        /**
         * getEntityData
         * @returns
         */
        getEntityData : function() {            
            return this.data;
        },
        
        /**
         * @method getSummary
         * @returns
         */
        getSummary : function() {
        	if (!this._summary && this._get("totalResults")[0]) {
                this._summary = {
                	totalResults : this.getAsNumber("totalResults"),
                	startIndex : this.getAsNumber("startIndex"),
                	itemsPerPage : this.getAsNumber("itemsPerPage")
                };
            }
            return this._summary;
        },
        
        /**
         * @method getEntitiesDataArray
         * @returns {Array}
         */
        getEntitiesDataArray : function() {
        	var entityJPath = this._getJPath("entry");
        	var resultingArray = this._get(entityJPath);
        	return resultingArray[0];
        },
        
        _getDate : function(property) {
        	var text = this._get(property)[0];
        	if(text instanceof Date) {
        		return text;
        	}
        	else {
        		return new Date(text);
        	}
        },
        
        _getBoolean : function(property) {
        	var text = this._get(property)[0];
        	return text ? true : false;
        },
        
        _getNumber : function(property) {
        	var text = this._get(property)[0];
        	// if it is a Number
        	if(typeof text === 'number')
        		return text;
        	//if its an array, we return the length of the array
        	else if(lang.isArray(text))
        		return text.length;
        	//if its a string or any other data type, we convert to number and return. Invalid data would throw an error here.
        	else return Number(text);
        },
        
        /**
         Validate that the property is valid
         **/
        _validateProperty : function(property, method) {
            if (!property) {
                var msg = stringUtil.substitute("Invalid argument for JsonDataHandler.{1} {0}", [ property, method ]);
                throw new Error(msg);
            }
        },
        /**
         Validate that the object is valid
         **/
        _validateObject : function(object) {
            if (!object) {
                var msg = stringUtil.substitute("Invalid argument for JsonDataHandler.{0}", [ object ]);
                throw new Error(msg);
            }
        },
        
		_evalJson: function(jsonQuery){
			return jsonPath(this.data,jsonQuery);			
		},
	
		_evalJsonWithData: function(jsonQuery, data){
			return jsonPath(data,jsonQuery);			
		},
		
		_get: function(property){

			this._validateObject(this.data);
			var jsonQuery = this._getJPath(property);

			var result = this._evalJson(jsonQuery);
			return result;
		},
		
		_getJPath: function(property) {
			return this.jsonpath[property] || property;
		},
		
		extractFirstElement: function(result){
			this._validateObject(result);
			return this._evalJsonWithData("$[0]", result);
		}
	});
	return JsonDataHandler;
});

File: sdk/com.ibm.sbt.core/src/main/java/com/ibm/sbt/services/client/base/datahandlers/JsonDataHandler.java
/*
 * © Copyright IBM Corp. 2013
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); 
 * you may not use this file except in compliance with the License. 
 * You may obtain a copy of the License at:
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, 
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or 
 * implied. See the License for the specific language governing 
 * permissions and limitations under the License.
 */

package com.ibm.sbt.services.client.base.datahandlers;

import java.text.DateFormat;
import java.text.ParseException;
import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;

import com.ibm.commons.util.io.json.JsonJavaObject;

/**
 * This class provides an implementation of the DataHandler class to use data in JSON format. 
 * @author Carlos Manias
 * @author Manish Kataria
 */
public class JsonDataHandler implements DataHandler<JsonJavaObject>{
	
	private JsonJavaObject data;

	/**
	 * Constructor
	 *  
	 * @param jsonObject
	 */
	public JsonDataHandler(JsonJavaObject jsonObject) {
		this.data = jsonObject;
	}
	
	/**
	 * @param field
	 * @return value 
	 */
	@Override
	public String getAsString(FieldEntry field) {
		return getAsString((String)field.getPath());
	}
	
	/**
	 * @param path
	 * @return value 
	 */
	@Override
	public String getAsString(String path) {
		return path.indexOf("/")==-1?data.getString(path):getNestedField(path);
	}
	
	/*
	 * @param path
	 * @return
	 */
	private String getNestedField(String path){
		String value = "";
		String[] pathParts = path.split("/");
		JsonJavaObject currentObject = data;
		int index = 0;
		for (String part : pathParts){
			if (!containsKey(currentObject,part)) break;
			if (index<pathParts.length-1) {
				currentObject = currentObject.getJsonObject(part);
			} else {
				value = currentObject.getString(part);
			}
			index++;
		}
		return value;
	}
	
	public boolean containsKey(JsonJavaObject j, String key) {
	    Iterator<String> i = j.getProperties();
	    while(i.hasNext()) {
	        if(i.next().matches(key))return true;
	    }
	    return false;
	}
	
	private Object getNestedObject(String path){
		String[] pathParts = path.split("/");
		JsonJavaObject currentObject = data;
		int index = 0;
		for (String part : pathParts){
			if (!containsKey(currentObject, part)) break;
			if (index<pathParts.length-1) {
				currentObject = currentObject.getJsonObject(part);
			}
			index++;
		}
		return currentObject;
	}
	
	@Override
	public JsonJavaObject getEntry(FieldEntry field) {
		return getEntry((String)field.getPath());
	}

	@Override
	public JsonJavaObject getEntry(String path) {
		return data.getJsonObject(path);
	}
	
	/**
	 * @param field
	 * @return list of entries
	 */
	@Override
	public List<JsonJavaObject> getEntries(FieldEntry field) {
		return getEntries((String)field.getPath());
	}
	
	/**
	 * @param path
	 * @return list of entries
	 */
	@Override
	public List<JsonJavaObject> getEntries(String path) {
		return getAsList(path);
	}
	
	/**
	 * @param field
	 * @return list of entries
	 */
	public List<JsonJavaObject> getAsList(FieldEntry field){
		return getAsList((String)field.getPath());
	}
	
	/**
	 * @param path
	 * @return list of entries
	 */
	public List<JsonJavaObject> getAsList(String path){
		List<JsonJavaObject> jsonList = new ArrayList<JsonJavaObject>();
		
		if(path.contains("/")){
			JsonJavaObject object = (JsonJavaObject) getNestedObject(path);
			String pathinfo = path.substring(path.lastIndexOf("/")+1,path.length());
			if(containsKey(object,pathinfo)){
				for (Object obj : (List)object.get(pathinfo)) {
					jsonList.add((JsonJavaObject)obj);
				}
			}else{
				return null;
			}
		}else{
			for (Object obj : (List)data.get(path)) {
				jsonList.add((JsonJavaObject)obj);
			}
		}
		return jsonList;
	}
	
	/* (non-Javadoc)
	 * @see com.ibm.sbt.services.client.base.datahandlers.DataHandler#getAsArray(com.ibm.sbt.services.client.base.datahandlers.FieldEntry)
	 */
	@Override
	public String[] getAsArray(FieldEntry field) {
		return getAsArray((String)field.getPath());
	}
	
	/* (non-Javadoc)
	 * @see com.ibm.sbt.services.client.base.datahandlers.DataHandler#getAsArray(java.lang.String)
	 */
	@Override
	public String[] getAsArray(String path) {
		List<String> strList = new ArrayList<String>();
		
		if(path.contains("/")){
			JsonJavaObject object = (JsonJavaObject) getNestedObject(path);
			String pathinfo = path.substring(path.lastIndexOf("/")+1,path.length());
			if(containsKey(object,pathinfo)){
				for (Object obj : (List)object.get(pathinfo)) {
					strList.add((String)obj);
				}
			}else{
				return null;
			}
		}else{
			for (Object obj : (List)data.get(path)) {
				strList.add((String)obj);
			}
		}
		return strList.toArray(new String[strList.size()]);
	}
	
	/**
	 * @set data
	 */
	@Override
	public void setData(final Object data) {
		this.data = (JsonJavaObject)data;
	}
	
	/**
	 * @return data
	 */
	@Override
	public JsonJavaObject getData() {
		return data;
	}

    /**
     * @param field
     * @param locale
     * @return value as Date 
     * @throws DataHandlerException 
     */
	@Override
	public Date getAsDate(FieldEntry field, final Locale locale) throws DataHandlerException {
		String value = getAsString(field);
		Date date = null;
		DateFormat format = DateFormat.getDateInstance(DateFormat.SHORT, locale);
		try {
			date = format.parse(value);
		} catch(ParseException e) {
			throw new DataHandlerException(e, "Error parsing date string");
		}
		return date;
	}
	
    /**
     * @param fieldName
     * @param locale
     * @return value as Date 
     * @throws DataHandlerException 
     */
	@Override
	public Date getAsDate(String fieldName, final Locale locale) throws DataHandlerException {
		String value = getAsString(fieldName);
		Date date = null;
		DateFormat format = DateFormat.getDateInstance(DateFormat.SHORT, locale);
		try {
			date = format.parse(value);
		} catch(ParseException e) {
			throw new DataHandlerException(e, "Error parsing date string");
		}
		return date;
	}
	
    /**
     * @param field
     * @return value as Date 
     * @throws DataHandlerException 
     */
	@Override
	public Date getAsDate(FieldEntry field) throws DataHandlerException {
		return getAsDate(field, Locale.getDefault());
	}
	
    /**
     * @param fieldName
     * @return value as Date 
     * @throws DataHandlerException 
     */
	@Override
	public Date getAsDate(String fieldName) throws DataHandlerException {
		return getAsDate(fieldName, Locale.getDefault());
	}

    /**
     * @param field
     * @return value as int 
     */
	@Override
	public int getAsInt(FieldEntry field) {
		return Integer.parseInt(getAsString(field));
//		return data.getAsInt((String)field.getPath());
	}

    /**
     * @param fieldName
     * @return value as int 
     */
	@Override
	public int getAsInt(String fieldName) {
		try {
			return Integer.parseInt(getAsString(fieldName));
		} catch (NumberFormatException e) {
			return 0;
		}
	}
	
	/**
     * @param fieldName
     * @return value as long 
     */
	@Override
	public Long getAsLong(String fieldName) {
		try {
			return Long.parseLong(getAsString(fieldName));
		} catch (NumberFormatException e) {
			return (long) 0;
		}
	}
	
	/**
     * @param fieldName
     * @return value as long 
     */
	@Override
	public Long getAsLong(FieldEntry fieldName) {
		String field = (String)fieldName.getPath();
		try {
			return Long.parseLong(getAsString(field));
		} catch (NumberFormatException e) {
			return (long) 0;
		}
	}

    /**
     * @param field
     * @return value as float 
     */
	@Override
	public float getAsFloat(FieldEntry field) {
		return (float)data.getDouble((String)field.getPath());
	}

    /**
     * @param fieldName
     * @return value as float 
     */
	@Override
	public float getAsFloat(String fieldName) {
		return (float)data.getDouble(fieldName);
	}

    /**
     * @param field
     * @return value as Date 
     */
	@Override
	public boolean getAsBoolean(FieldEntry field) {
		return data.getBoolean((String)field.getPath());
	}

    /**
     * @param fieldName
     * @return value as boolean 
     */
	@Override
	public boolean getAsBoolean(String fieldName) {
		return data.getBoolean(fieldName);
	}
}
